FuelLabs/fuelup
# feat(fuel-toolchain.toml): support dates

closes #331

~Marked this as ready prematurely but converting back to a draft for now while I work on possibly refactoring parts of the `Toolchain` and `Channel` abstractions. There are various overlapping concepts here as a result of mostly forking the functionality of `rustup`, going to leave my thoughts below:~

~We currently have `DistToolchainDescription` to reason about toolchains - it usually comes in the form of `<name>-<date>-<target>`, with `date` and `target` being optional. It is used as a way to parse user input. It is valid for users to do `fuelup toolchain install latest`, and also `fuelup toolchain install nightly-2023-01-10`.~

~Currently `latest` has no knowledge or concept of dates - so this has to be updated. This includes various places of the codebase.~

EDIT:

This should be ready for review.

## Overview

We're using `DistToolchainDescription` whereever possible to display and parse toolchain inputs; because of this abstraction thankfully we're easily able to enforce dated `latest`s with this PR. The concepts surrounding `DistToolchainDescription` do seem a bit messy atm and I think this should call for some minor refactoring in future, but I'll leave that for a later PR.

## How to test?

First put this convenience script `build.sh` in fuelup dir:

```sh
cargo build
rm -f ~/.fuelup/bin/forc
rm -f ~/.fuelup/bin/fuelup
ln target/debug/fuelup ~/.fuelup/bin/forc
ln target/debug/fuelup ~/.fuelup/bin/fuelup
```

Use this to run cargo build and it should automatically link dev fuelup that handles the new logic from this PR. Note that this messes up your actual `forc` and `fuelup` so you would have to reinstall fuelup after you're done reviewing.

Create a `fuel-toolchain.toml`, declaring `latest-2023-01-06`:

```toml
[toolchain]
channel = "latest-2023-01-06"
```

This channel was manually deployed by me (we don't have any new versions yet, but I used the date of the last release which was forc v0.33)

Assuming you've ran the above `build.sh`, now try running `forc`. If you don't have the above toolchain installed, `fuelup` will now fetch and install it for you.

Now try it again with a random dated nightly:

```toml
[toolchain]
channel = "nightly-2022-09-23"
```

Now try using `latest` and `nightly` (without dates), this should warn you that the toml is invalid, and default to using your default toolchain's `forc`.
